Source/Search/Mixins/AdvancedSearchProviderMixin.lua

```diff
@@ -0,0 +1,225 @@
+    maxLevel = nil
+  end
+
+  if minItemLevel == 0 then
+    minItemLevel = nil
+  end
+
+  if maxItemLevel == 0 then
+    maxItemLevel = nil
+  end
+
+  local exactSearch = ExtractExactSearch(queryString)
+
+  return {
+    query = {
+      searchString = queryString,
+      minLevel = minLevel,
+      maxLevel = maxLevel,
+      filters = {},
+      itemClassFilters = itemClassFilters,
+      sorts = {},
+    },
+    extraFilters = {
+      minItemLevel = minItemLevel,
+      maxItemLevel = maxItemLevel,
+      exactSearch = exactSearch,
+    }
+  }
+end
+
+function AuctionatorAdvancedSearchProviderMixin:CreateSearchTerm(term)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:CreateSearchTerm()", term)
+  if IsCompoundSearch(term) then
+    return ParseCompoundSearch(term)
+  else
+    return  {
+      query = {
+        searchString = term,
+        filters = {},
+        itemClassFilters = {},
+        sorts = {},
+      },
+      extraFilters = {
+        exactSearch = ExtractExactSearch(term)
+      }
+    }
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:GetSearchProvider()
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:GetSearchProvider()")
+
+  --Run the query, and save extra filter data for processing
+  return function(searchTerm)
+    C_AuctionHouse.SendBrowseQuery(searchTerm.query)
+    self.currentFilter = searchTerm.extraFilters
+    self.itemKeyInfoQueue = {}
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:HasCompleteTermResults()
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:HasCompleteTermResults()")
+
+  --Loaded all the terms from API, and we have filtered every item
+  return C_AuctionHouse.HasFullBrowseResults() and #(self.itemKeyInfoQueue) == 0
+end
+
+function AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived(eventName, ...)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived()", eventName, ...)
+
+  if eventName == "AUCTION_HOUSE_BROWSE_RESULTS_UPDATED" then
+    self:ProcessSearchResults(C_AuctionHouse.GetBrowseResults())
+  elseif eventName == "AUCTION_HOUSE_BROWSE_RESULTS_ADDED" then
+    self:ProcessSearchResults(...)
+  elseif eventName == "ITEM_KEY_ITEM_INFO_RECEIVED" then
+    self:ProcessItemKeyInfo(...)
+  elseif eventName == "AUCTION_HOUSE_BROWSE_FAILURE" then
+    AuctionHouseFrame.BrowseResultsFrame.ItemList:SetCustomError(
+      RED_FONT_COLOR:WrapTextInColorCode(ERR_AUCTION_DATABASE_ERROR)
+    )
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults(addedResults)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults()")
+
+  local results = {}
+
+  for index = 1, #addedResults do
+    -- Run filter checks on every item key. Some might not be added to the
+    -- results yet, but when the relevant information arrives in an event
+    if self:FilterByItemLevel(addedResults[index].itemKey) then
```

Comment:
Combine nested boolean expression

Reply:
The logic for the filters relied on the short circuit property, I wasn't sure if it counted as "too smart" to group them

Reply:
I'm OK with short-circuiting logic because it reads linearly, so its easy to understand (especially if the boolean expression is fairly straightforward like all `or`s or `and`s). At the point that something becomes "too smart", I tend to reach for functions as a tool to increase clarity and to make code more concise by providing meaningful names around chunks of logic.